cli/internal/encoding/lstree/lstree.go

```diff
@@ -0,0 +1,275 @@
+	ErrFieldCount        = errors.New("too many fields")
+)
+
+// A Reader reads records from `git ls-tree`'s output`. The Reader converts
+// all \r\n sequences in its input to plain \n.
+type Reader struct {
+	// ReuseRecord controls whether calls to Read may return a slice sharing
+	// the backing array of the previous call's returned slice for performance.
+	// By default, each call to Read returns newly allocated memory owned by the caller.
+	ReuseRecord bool
+
+	reader *bufio.Reader
+
+	// numEntry is the current entry being read in the ls-tree output.
+	numEntry int
+
+	// rawBuffer is an entry buffer only used by the readEntry method.
+	rawBuffer []byte
+
+	// recordBuffer holds the unescaped fields, one after another.
+	// The fields can be accessed by using the indexes in fieldIndexes.
+	// E.g., For the row `a,"b","c""d",e`, recordBuffer will contain `abc"de`
+	// and fieldIndexes will contain the indexes [1, 2, 5, 6].
+	recordBuffer []byte
+
+	// fieldIndexes is an index of fields inside recordBuffer.
+	// The i'th field ends at offset fieldIndexes[i] in recordBuffer.
+	fieldIndexes []int
+
+	// fieldPositions is an index of field positions for the
+	// last record returned by Read.
+	fieldPositions []position
+
+	// lastRecord is a record cache and only used when ReuseRecord == true.
+	lastRecord []string
+}
+
+// NewReader returns a new Reader that reads from r.
+func NewReader(reader io.Reader) *Reader {
+	return &Reader{
+		reader: bufio.NewReader(reader),
+	}
+}
+
+// Read reads one record from `reader`.
+// Read always returns either a non-nil record or a non-nil error,
+// but not both.
+//
+// If there is no data left to be read, Read returns nil, io.EOF.
+//
+// If ReuseRecord is true, the returned slice may be shared
+// between multiple calls to Read.
+func (r *Reader) Read() (record []string, err error) {
+	if r.ReuseRecord {
+		record, err = r.readRecord(r.lastRecord)
+		r.lastRecord = record
+	} else {
+		record, err = r.readRecord(nil)
+	}
+	return record, err
+}
+
+// FieldPos returns the entry and column corresponding to
+// the start of the field with the given index in the slice most recently
+// returned by Read. Numbering of entries and columns starts at 1;
+// columns are counted in bytes, not runes.
+//
+// If this is called with an out-of-bounds index, it panics.
+func (r *Reader) FieldPos(field int) (entry int, column int) {
+	if field < 0 || field >= len(r.fieldPositions) {
+		panic("out of range index passed to FieldPos")
+	}
+	p := &r.fieldPositions[field]
+	return p.entry, p.col
+}
+
+// pos holds the position of a field in the current entry.
+type position struct {
+	entry, col int
+}
+
+// ReadAll reads all the records from reader until EOF.
+//
+// A successful call returns err == nil, not err == io.EOF. Because ReadAll is
+// defined to read until EOF, it does not treat end of file as an error to be
+// reported.
+func (r *Reader) ReadAll() (records [][]string, err error) {
```

Comment:
Is this return type fixed by some interface? Or can we do better with a `struct`, since this is specific to `ls-tree`? If we could get `entry.File` and `entry.Hash` instead of `entry[1]` and `entry[2]` or whatever, I think that would be nice

Reply:
This definitely isn't a fixed interface this is just how the CSV reader did it. They convert to a string once to pin the memory together and then slice it at the indices. Since this was perf-focused Go code and I don't yet know the consequences of changing the shape on perf I decided to keep the API as it was. 🤷‍♂️

Right now it is `ls-tree`-specific but I plan to generalize it for `ls-files` and `status`. All of the `git <command> -z` serializations are `<field> <field> <field>\t<path>\000`. So I'll just define a few constants specifying the field names and length. That doesn't preclude putting it into a struct; I'd just do it as a "last pass" over the tuple, keeping this API at the core and hiding it away.